src/Commands/ElasticsearchIndexCommand.php

```diff
@@ -4,55 +4,255 @@
+    {
+        $this->progressBar = false;
+
+        return $this;
+    }
+
+    public function withSynonymsFor(array $for)
+    {
+        if (count($for)) {
+            $this->synonymsFor = array_merge($this->synonymsFor, $for);
+        }
+
+        return $this;
     }
 
-    public function indexAllStores(string $indexName, callable|iterable $items, callable|array|null $dataFilter, callable|string $id = 'entity_id'): void
+    public function withExtraSynonyms(array $synonyms)
     {
-        $this->indexStores(Rapidez::getStores(), $indexName, $items, $dataFilter, $id);
+        if (count($synonyms)) {
+            $this->extraSynonyms = array_merge($this->extraSynonyms, $synonyms);
+        }
+
+        return $this;
     }
 
-    public function indexStores(array $stores, string $indexName, callable|iterable $items, callable|array|null $dataFilter, callable|string $id = 'entity_id'): void
+    public function index(string $indexName, callable|iterable|Builder $items, callable|string $id = 'entity_id'): void
     {
-        foreach ($stores as $store) {
-            $this->indexStore($store, $indexName, $items, $dataFilter, $id);
+        foreach ($this->stores as $store) {
+            $this->indexStore(
+                store: $store,
+                indexName: $indexName,
+                items: $items,
+                id: $id,
+            );
         }
     }
 
-    public function indexStore(Store|array $store, string $indexName, callable|iterable $items, callable|array $dataFilter, callable|string $id = 'entity_id'): void
+    public function initIndex(Store|array $store, string $indexName)
     {
         $storeName = $store['name'] ?? $store['code'] ?? reset($store);
         $this->line('Indexing `' . $indexName . '` for store ' . $storeName);
+        $this->prepareIndexerWithStore($store, $indexName);
+    }
+
+    public function beforeIndex(int $totalCount)
+    {
+        if ($this->progressBar) {
+            $this->bar = $this->output->createProgressBar($totalCount);
+            $this->bar->start();
+        }
+    }
+
+    public function afterIndex()
+    {
+        $this->indexer->finish();
+        if ($this->progressBar) {
+            $this->bar->finish();
+            $this->line('');
+        }
+    }
 
+    public function indexStore(Store|array $store, string $indexName, callable|iterable|Builder $items, callable|string $id): void
+    {
+        $this->initIndex($store, $indexName);
+        $fullItems = $this->dataFrom($items);
+
+        if (is_null($fullItems)) {
+            $this->indexer->abort();
+
+            throw new Exception('Items cannot be null.');
+        }
+
+        $count = is_iterable($fullItems) ? count($fullItems) : $fullItems->getQuery()->getCountForPagination();
+
+        $this->beforeIndex($count);
+        $this->tryIndexItems($fullItems, $id);
+        $this->afterIndex();
+    }
+
+    public function tryIndexItems(iterable|Builder $items, callable|string $id): void
```

Comment:
I think these functions would be better suited in a class managing elasticsearch instead of an abstract command

Reply:
I agree with this, however the problem I have with doing that is that the progress bar logic (coming all the way from `indexPartial`) would then be moved out of the command which doesn't make sense to me.

I might be able to add some callback parameter to advance the bar? Doesn't feel very clean to me either. Could also just get rid of the bar entirely but it's nice to have.

Reply:
You're right in that when using this in the current way. However when using a lazy collection we can cheat a little bit by using the map method on the lazy collection.

The map method on a Lazy collection is different than that of a regular collection. 
It will get executed once the data gets accessed, instead of being executed once called.

So each iteration of the foreach loop would call the map function

Reply:
Ah, a LazyCollection is just essentially a generator wrapper I see. Should make it a lot easier to get a clean solution here 🤔